packages/core/crates/core-network/src/ping.rs

```diff
@@ -0,0 +1,512 @@
+                "Measures total time it takes to probe all other nodes (in seconds)",
+                vec![0.5, 1.0, 2.5, 5.0, 10.0, 15.0, 30.0, 60.0, 90.0, 120.0, 300.0]
+            ).ok(),
+            metric_time_to_ping: SimpleHistogram::new(
+                "core_histogram_ping_time_seconds",
+                "Measures total time it takes to ping a single node (seconds)",
+                vec![0.5, 1.0, 2.5, 5.0, 10.0, 15.0, 30.0, 60.0, 90.0, 120.0, 300.0]
+            ).ok(),
+            metric_successful_ping_count: SimpleCounter::new(
+                "core_counter_heartbeat_successful_pings",
+                "Total number of successful pings"
+            ).ok(),
+            metric_failed_ping_count: SimpleCounter::new(
+                "core_counter_heartbeat_failed_pings",
+                "Total number of failed pings"
+            ).ok(),
+        }
+    }
+
+    pub async fn ping_peers<F>(&self, mut peers: Vec<PeerId>,
+                                 send_msg: &impl Fn(Box<[u8]>, String) -> F)
+    where
+        F: futures::Future<Output =  Result<Box<[u8]>, String>>
+    {
+        if peers.is_empty() {
+            // debug!("Received an empty peer list");
+            ()
+        }
+
+        let _ping_peers_timer = match &self.metric_time_to_heartbeat {
+            Some(metric_time_to_heartbeat) => {
+                let timer = metric_time_to_heartbeat.start_measure();
+                Some(scopeguard::guard((), move |_| { metric_time_to_heartbeat.cancel_measure(timer); }))
+            },
+            None => None
+        };
+
+        let remainder = peers.split_off(self.config.max_parallel_pings.min(peers.len()));
+
+        let start = current_timestamp();
+        let mut futs = to_futures_unordered(
+            peers.iter()
+                .map(|x| { self.ping_peer(x.clone(), self.config.timeout, send_msg)})
+                .collect::<Vec<_>>()
+        );
+
+        let mut waiting = remainder.iter();
+        while let Some(heartbeat) = futs.next().await {
+            if let Some(v) = waiting.next() {
+                let remaining_time = current_timestamp() - start;
+                if (remaining_time as u128) < self.config.timeout.as_millis() {
+                    futs.push(self.ping_peer(v.clone(), Duration::from_millis(remaining_time), send_msg));
+                }
+            }
+
+            self.external_api.on_finished_ping(&heartbeat.0, heartbeat.1);
+        }
+    }
+
+    async fn ping_peer<F>(&self, destination: PeerId, timeout_duration: Duration,
+                       send_msg: &impl Fn(Box<[u8]>, String) -> F) -> PingMeasurement
+    where
+        F: futures::Future<Output = Result<Box<[u8]>, String>>
+    {
+        // TODO: log(format!("Pinging the peer '{}'", destination));
+        use rand::{RngCore,SeedableRng};
+        let mut challenge = Box::new([0u8; 16]);
+        rand::rngs::StdRng::from_entropy().fill_bytes(&mut challenge.as_mut_slice());
```

Comment:
I was thinking about this too, but to make the ping stateless, the RNG has to be instantiated inside the ping. Relative to a ping operation the instantiation is likely fast.

Reply:
So what is the problem with statefulness? Well, the transport cannot be passed into the future. And that's where the entire model falls apart. Having the generator as a static generator, but then the randomness would not be guaranteed on a multi-threaded executor.

Reply:
Let's document this as a tech debt and we'll get back to it, because we'll need to think about safe (and performant) randomness certainly again in the near future.

Reply:
Added to https://github.com/hoprnet/hoprnet/issues/4729